src/main/kotlin/org/rust/cargo/runconfig/target/RsCommandLineSetup.kt

```diff
@@ -0,0 +1,117 @@
+    private val environmentPromise = AsyncPromise<Pair<TargetEnvironment, TargetProgressIndicator>>()
+    private val dependingOnEnvironmentPromise: MutableList<Promise<Unit>> = mutableListOf()
+    private val uploads: MutableList<Upload> = mutableListOf()
+    private val projectHomeOnTarget: VolumeDescriptor = VolumeDescriptor(
+        RsCommandLineSetup::class.java.simpleName + ":projectHomeOnTarget",
+        "",
+        "",
+        "",
+        request.projectPathOnTarget
+    )
+
+    fun requestUploadIntoTarget(uploadPathString: String): TargetValue<String> {
+        val uploadPath = FileUtil.toSystemDependentName(uploadPathString).toPath()
+        val isDir = uploadPath.isDirectory()
+        val localRootPath = if (isDir) uploadPath else (uploadPath.parent ?: Paths.get("."))
+        val (uploadRoot, pathToRoot) = request.getUploadRootForLocalPath(localRootPath.toString())
+            ?: createUploadRoot(projectHomeOnTarget, localRootPath).let { uploadRoot ->
+                request.uploadVolumes += uploadRoot
+                uploadRoot to "."
+            }
+        val result = DeferredTargetValue(uploadPathString)
+        dependingOnEnvironmentPromise += environmentPromise.then { (environment, targetProgressIndicator) ->
+            if (targetProgressIndicator.isCanceled || targetProgressIndicator.isStopped) {
+                result.stopProceeding()
+                return@then
+            }
+            val volume = environment.uploadVolumes.getValue(uploadRoot)
+            try {
+                val relativePath = if (isDir) pathToRoot
+                else uploadPath.fileName.toString().let {
+                    if (pathToRoot == ".") it else joinPath(pathToRoot, it)
+                }
+                val resolvedTargetPath = volume.resolveTargetPath(relativePath)
+                uploads.add(Upload(volume, relativePath))
+                result.resolve(resolvedTargetPath)
+            } catch (t: Throwable) {
+                LOG.warn(t)
+                targetProgressIndicator.stopWithErrorMessage(
+                    LangBundle.message(
+                        "progress.message.failed.to.resolve.0.1",
+                        volume.localRoot, t.localizedMessage
+                    )
+                )
+                result.resolveFailure(t)
+            }
+        }
+        return result
+    }
+
+    private fun createUploadRoot(
+        volumeDescriptor: VolumeDescriptor,
+        localRootPath: Path
+    ): TargetEnvironment.UploadRoot =
+        languageRuntime?.createUploadRoot(volumeDescriptor, localRootPath)
+            ?: TargetEnvironment.UploadRoot(localRootPath, Temporary())
+
+    private fun joinPath(vararg segments: String): String =
+        segments.joinToString(request.targetPlatform.platform.fileSeparator.toString())
+
+    fun provideEnvironment(environment: TargetEnvironment, targetProgressIndicator: TargetProgressIndicator) {
```

Comment:
As I see, a large part of `RsCommandLineSetup` implementation, and particularly this method as an example, is mostly copied from implementations for other languages in IntelliJ repo. I understand that at the moment we have to copy-paste it, but in the future, it will cost us time to fetch and adopt the ongoing changes. 

Not sure about the stability of the Run targets API, but it would be great to consider this potential problem and maybe initiate necessary refactorings in IntellIJ Platform to reuse the code without copy-pasting.

Cc @Undin

Reply:
The Run targets API is not stable and as I heard it would be simplified in the near future, so I think the current solution is OK.